deeplay/external/layer.py

```diff
@@ -1,13 +1,95 @@
-from typing import Any, Callable, Type, overload
+from typing import Any, Type, overload, Dict, Union, List
 from .external import External
+from functools import partial
 
 import torch.nn as nn
 
+from ..decorators import after_build
+
+
+def _create_forward_with_input_dict(
+    old_forward,
+    input_arguments: List[str],
+    input_kwargs: Dict[str, str],
+    output_arguments: Dict[str, int],
+):
+    def forward_with_input_dict(self, x):
+        x = x.copy()
+
+        outputs = old_forward(
+            self,
+            *[x.get(arg) for arg in input_arguments],
+            **{key: x.get(value) for key, value in input_kwargs.items()},
+        )
+
+        if not isinstance(outputs, tuple):
+            outputs = (outputs,)
+
+        expected_outputs = len(set(output_arguments.values()))
+        assert len(outputs) == expected_outputs, (
+            f"layer {self} returned {len(outputs)} outputs, "
+            f"but it should return {expected_outputs}"
+        )
+
+        x.update({key: outputs[value] for key, value in output_arguments.items()})
+        return x
+
+    return forward_with_input_dict
+
 
 class Layer(External):
     def __pre_init__(self, classtype: Type[nn.Module], *args, **kwargs):
         super().__pre_init__(classtype, *args, **kwargs)
 
+    def set_dict_input_mapping(self, *args: str, **kwargs: str):
+        self.__dict__.update(
+            {
+                "input_arguments": args,
+                "input_kwargs": kwargs,
+                "_dict_input_mapping_called": True,
+            }
+        )
+        self._check_and_execute_dict_mapping()
+
+    def set_dict_output_mapping(self, *args: str, **kwargs: int):
```

Comment:
I'm not sure the flexibility to provude the output mapping as kwargs makes sense. I can't think of any case where you would not be able to achieve what you want with just `*args`

Reply:
The intention behind allowing the output mapping as `**kwargs` was to accommodate scenarios where you wish to allocate outputs to multiple targets, such as `external.set_dict_output_mapping("x", other_x_1=0, other_x_2=0)`. Can you suggest an alternative using `*args` exclusively?

Reply:
I see what you mean. Makes sense. But, I think I prefer the syntax of ("x", ("out_1", "out_2")).

Reply:
@giovannivolpe @Henrik-KM @cmanzo any thoughts on this?